include/zephyr/drivers/video.h

```diff
@@ -118,6 +118,70 @@ struct video_buffer {
 	uint32_t timestamp;
 };
 
+/**
+ * @brief video_frmival_type enum
+ *
+ * Supported frame interval type of a video device.
+ */
+enum video_frmival_type {
+	/** discrete frame interval type */
+	VIDEO_FRMIVAL_TYPE_DISCRETE = 1,
+	/** stepwise frame interval type */
+	VIDEO_FRMIVAL_TYPE_STEPWISE = 2,
+};
+
+/**
+ * @struct video_frmival
+ * @brief Video frame interval structure
+ *
+ * Used to describe a video frame interval.
+ */
+struct video_frmival {
+	/** numerator of the frame interval */
+	uint32_t numerator;
+	/** denominator of the frame interval */
+	uint32_t denominator;
+};
+
+/**
+ * @struct video_frmival_stepwise
+ * @brief Video frame interval stepwise structure
+ *
+ * Used to describe the video frame interval stepwise type.
+ */
+struct video_frmival_stepwise {
+	/** minimum frame interval in seconds */
+	struct video_frmival min;
+	/** maximum frame interval in seconds */
+	struct video_frmival max;
+	/** frame interval step size in seconds */
+	struct video_frmival step;
+};
+
+/**
+ * @struct video_frmival_enum
+ * @brief Video frame interval enumeration structure
+ *
+ * Used to describe the supported video frame intervals of a given format.
+ */
+struct video_frmival_enum {
+	/** frame interval index during enumeration */
```

Comment:
This seems to include what we already describe in video_format struct, why not using/passing that struct instead.

Reply:
@loicpoulain I 'd like to do so but unfortunately, the video_format struct contains the **_pitch field_** which is not used here. Unused fields waste memory space especially when it is in an API definition as many drivers will implement it.

Otherwise, I think the pitch field is not needed in the video_format struct as it can be computed separately from pixel format and width field. If you feel agree with this, I can submit a PR to remove this field from video_format then using the video_format struct here ?

Reply:
Hmm AFAIU, pitch can depend on the capture pipeline, and depend also on some alignment requirements, so not sure it's something we can obtain only from pixeld format and width.

When we want to enumerate the allowed frame intervals, we normally already picked a frame format (e.g video_get_format), so it's odd to me we need to copy it's value back for enumerating frame intervals, instead of just passing a reference to the current format.

Reply:
Yes, I agree that pitch in the simplest case is` width * bpp` but in general it depends on other factors such as alignment.

BTW, for a clean code, especially for an API definition, I think we should not pass something that is more than needed, AFAIK in upstream Linux. It will create a bad precedent if we do it.

Reply:
@loicpoulain: Replacing :

```
struct video_frmival_enum {
	...
	uint32_t pixelformat;
	uint32_t width;
	uint32_t height;
	...
}
```
by 

```
struct video_frmival_enum {
	...
	struct video_format;
	...
}
```

where

```
struct video_format {
	uint32_t pixelformat;
	uint32_t width;
	uint32_t height;

	**_uint32_t pitch;_**
};
```

would be convenient if we already have the current video_format struct, so we just pass it to the function.
But if we want to query the supported frame rates for other video formats, we still have to construct the video_format struct anyways and leaving a never used field (e.g. pitch) in the struct will make developers confused as they don't know why they need or how to fill this pitch field, I think.

Other drawbacks is that never-used fields will waste memory space and could make a "dangerous point" (imagine that this field will be sent from application to the drivers without any sanity checks ...).

But well, if you see it's more convenient this way, I am going to change it.

Reply:
Maybe the pitch does have some usefulness...
It introduces the idea of padding, and the padding data, while not useful, affects the frame rate too?

For instance, these two rows of pixels would surely take different amount of time to transmit:

```
[  frame data  |  padding              ] <- larger pitch,same width
```
```
[  frame data  |  padding  ] <- smaller pitch, same width
```

That sounds like some rare use-case though, at least for uncompressed frame images.

Reply:
Now that `struct video_format` is there, would it work to pass it separately?

Something like this:
```diff
 static int video_xyzzy_enum_frmival(const struct device *dev, enum video_endpoint_id ep,
+				     const struct video_format *query,
				     struct video_frmival_enum *response);
```

Then the application could have a const (flash) array of video formats it supports, and try to pass them one by one?

```c
for (int i = 0; i < ARRAY_SIZE(supported_fmts); i++) {
	struct video_frmival_enum response;
	struct video_frmival frmival;

	video_enum_frmival(video_dev, VIDEO_EP_OUT, &supported_fmts[i], &response);
	if (search_acceptable_frame_interval(&response, &frmival) == 0) {
		return video_set_frmival(video_dev, VIDEO_EP_OUT, &frmival);
	}
}
return -ENOTSUP;
```

Maybe this is what @loicpoulain meant by "passing a reference".

Reply:
@josuah : Sorry I was a bit late on responding ... Yes, I understood what @loicpoulain meant is to pass a `video_format` struct as a reference to the `enum_frmival()` API instead of passing 3 separate parameters (width, height, pixel_format). It's more convenient to do that as the `video_format `struct is already defined.

The point being discussed is the `video_format` struct contains a pitch field which we will never use. You are right that the `pitch ` does affect the pixel rate / frame rate (although it is usually small and neglectable). But the main reason here is to query of supported frame rates, we don't need to give the pitch. We ask the pipeline :

- Give me the frame rates supported for this `{width, height, pixel_format}`

The pipeline (e.g. : ov5640 -> mipi csi2rx -> csi) will know to answer with only these parameters as input. The sensor driver itself already knows which frame rates it can support for 720p RGB565 format for example, **if there are any padding needed, it already took them into account internally**. Then, the mipi csi2rx will base on the required pixel rate or link frequency from the sensor (see https://github.com/zephyrproject-rtos/zephyr/pull/76475/commits/e3ad02203ed5f57874071e735c635efa7b57065c#diff-f7423f09c35fbf23beefd1f706f4e1dbd5477daf2c40fbc1d6b731d206156bd9R239) to derive the supported frame rates of the whole pipeline.

It's like when you ask a carpenter "Give me the price of a wooden bookcase of size 1m x 0.5m x 2m", you don't need to precise "you should take a little bit more wood than 1x 0.5 x 2" to compensate for the loss during the production".

Reply:
Good point, no need to explain to a device what is its own pitch: it knows about it already.

Reply:
@loicpoulain : Replacing `{width, height, pixel_format}` by `struct video_format` as your request (the pitch field is never used but well it's not a big deal).